### PR TITLE
Benjamin rogers

### PR DESCRIPTION
**Pull request by thisbenrogers**
@ squashgray couldn't add you as a reviewer yet.

### CODE REVIEW

File: src/03_modules.py [hunk 1]
```diff
@@ -10,22 +10,30 @@
...
+print(f"{sys.version_info[0]}.{sys.version_info[2]}.{sys.version_info[2]} {sys.version_info[3]}")
 
 
 import os
 # See the docs for the OS module: https://docs.python.org/3.7/library/os.html
 
 # Print the current process ID
 # YOUR CODE HERE
+pid = os.getpid()
+print(os.getpgid(pid))
```
Comment (birdcar) [inline src/03_modules.py:30]: Technically speaking, this provides the group process ID and not the process ID itself. These are at times the same, but not always (for example, when the current process was spawned by another process rather than directly).

Your first line, the `os.getpid()` is the most consistent way to get a process ID (assuming of course that you want the process that is running this file specifically, and not the process that might have spawned the process that is running the file).

File: src/05_lists.py [hunk 2]
```diff
@@ -8,22 +8,29 @@
...
 # Print the length of list x
 # YOUR CODE HERE
+print(len(x))
 
 # Print all the values in x multiplied by 1000
-# YOUR CODE HERE
\ No newline at end of file
+# YOUR CODE HERE
+times = [i * 1000 for i in x]
+print(times)
```
Comment (birdcar) [inline src/05_lists.py:36]: This is not a problem in any way, but if you don't _need_ to save the list you generated into memory, you could actually just pass the list comprehension directly to the `print` function.

```python
print([i * 1000 for i in x])
```

File: src/08_comprehensions.py [hunk 3]
```diff
@@ -10,14 +10,14 @@
 
 # Write a list comprehension to produce the array [1, 2, 3, 4, 5]
 
-y = []
+y = [x for x in range(6)]
```
Comment (birdcar) [inline src/08_comprehensions.py]: This actually produces 0-5 rather than 1-5.

```python
>>> [x for x in range(6)]                                                  
[0, 1, 2, 3, 4, 5]
```

To fix this you could do a few things, some examples would be:

1. Add a ["start" argument to the `range`](https://docs.python.org/3/library/functions.html#func-range) function: `[x for x in range(1, 6)]`
2. Add a filter to the comprehension: `[x for x in range(6) if x != 0]` or `[x for x in range(6) if x]`

Of these, the first would likely be the most Pythonic.

File: src/08_comprehensions.py [hunk 4]
```diff
@@ -36,6 +36,6 @@
 x = input("Enter comma-separated numbers: ").split(',')
 
 # What do you need between the square brackets to make it work?
-y = []
+y = [int(z) for z in x if int(z) % 2 == 0]
```
Comment (birdcar) [inline src/08_comprehensions.py]: Your answer is absolutely correct, but let's have some fun 😄 

Here are other ways we might do this.

First, since 0 is Falsey and all other numbers are Truthy, we could actually get rid of the `==0` part and write the comprehension like so:

```python
[int(z) for z in x if not int(z) % 2]
```

A second fun thing we could do is change the input function to make it automatically parse the input numbers as integers. Which saves us two `int` calls in the list comprehension. 

We could do this using [Python's `map` function](https://docs.python.org/3/library/functions.html#map), which takes a function reference as the first argument (uncalled) and an iterable as the second argument (list, tuple, string, etc).

```python
x = map(int, input("Enter comma-separated numbers: ").split(','))
```

The thing about map is that it returns a `map` generator object. It waits to actually process the data until we consume it somehow. The good news is that using it in a list comprehension is one way we can consume it. So now, our code looks like this in total:

```python
x = map(int, input("Enter comma-separated numbers: ").split(','))

# What do you need between the square brackets to make it work?
y = [z for z in x if not z % 2]  

print(y)
```

The benefit of this process is that we don't actually loop over the list and make integers out of them until we need to, so we only ever loop over the list passed in by the user once, and we only ever have to create one list in memory.

File: src/09_dictionaries.py [hunk 5]
```diff
@@ -35,13 +35,23 @@
 
 # Add a new waypoint to the list
 # YOUR CODE HERE
+new_dict = {
+    "lat": 22,
+    "lon": -22,
+    "name": "The new place"
+}
+
+waypoints.append(new_dict)
```
Comment (birdcar) [inline src/09_dictionaries.py]: Rather than creating a dict named `new_dict` in memory and then appending that dictionary to `waypoints`, you could actually just append that directly like so:

```python
waypoints.append({ 
    "lat": 22, 
    "lon": -22, 
    "name": "The new place" 
})   
```

File: src/09_dictionaries.py [hunk 6]
```diff
@@ -35,13 +35,23 @@
...
 
 # Modify the dictionary with name "a place" such that its longitude
 # value is -130 and change its name to "not a real place"
 # Note: It's okay to access the dictionary using bracket notation on the
 # waypoints list.
 
 # YOUR CODE HERE
-
+for d in waypoints:
+    d.update((k, "not a real place") for k, v in d.items() if v == "a place")
```
Comment (birdcar) [inline src/09_dictionaries.py:53]: This is _excellent_ ✨

File: src/09_dictionaries.py [hunk 7]
```diff
@@ -35,13 +35,23 @@
...
 # Write a loop that prints out all the field values for all the waypoints
-# YOUR CODE HERE
\ No newline at end of file
+# YOUR CODE HERE
+for d in waypoints:
+    print(list(d.values()))
```
Comment (birdcar) [inline src/09_dictionaries.py:58]: This is absolutely correct, however, we could be slightly more time and space efficient.

In this writing, the `list` function first iterates over all of the values in `d.values()` and generates a list, and then it passes that list to print. 

We actually don't need to do that since the `values` method will create an iterable for us.

In this case, we could use the splat operator (`*`) to tell Python we want each item produced by the `values` method passed to the `print` function as a separate argument.

```python
for d in waypoints:
    print(*d.values())
```

If we wanted to be fancy, we could also tell `print` to put each value on its own line by passing `print` a `sep` keyword argument:

```python
for d in waypoints:
    print(*d.values(), sep="\n") 
```

File: src/10_functions.py [hunk 8]
```diff
@@ -1,6 +1,11 @@
 # Write a function is_even that will return true if the passed-in number is even.
 
 # YOUR CODE HERE
+def is_even(n):
+    if n % 2 == 0:
+        return True
+    else:
+        return False
```
Comment (birdcar) [inline src/10_functions.py]: This is my _absolute favorite_ code review comment, because it bites me all the time.

Comparison operators (i.e. `<`, `>`, `<=`, `>=`, `==`) _always_ produce a boolean value. The only exception to that is when they raise an error because you're comparing two objects that can't be compared.

With that in mind, we can actually remove both the `if` and `else` block entirely and just return the comparison:

```python
def is_even(n):
    return n % 2 == 0
```

You could also shorten it, but arguably make it less readable, to this:

```python
def is_even(n):
    return not n % 2
```

Which will return `True` every time `n % 2` results in `0`, and `False` otherwise.

File: src/10_functions.py [hunk 9]
```diff
@@ -9,4 +14,7 @@
 # Print out "Even!" if the number is even. Otherwise print "Odd"
 
 # YOUR CODE HERE
-
+if is_even(num) == True:
+    print("Even!")
+else:
+    print("Odd")
```
Comment (birdcar) [inline src/10_functions.py:17]: Since we know that `is_even` will always return a boolean, the `== True` part of the expression is redundant. As a result, we can shorten it to this:

```python
if is_even(num):
    print("Even!")
else:
    print("Odd")
```

Since each part of the `if`/`else` conditional only has one statement in it, you could also shorten this to a turnery expression like so:

```python
print("Even!") if is_even(num) else print("Odd")
```

☝️ this would be the equivalent of the JS code:

```javascript
isEven(num) ? console.log('Even!') : console.log('Odd');
```

File: src/11_args.py [hunk 10]
```diff
@@ -13,6 +15,11 @@
 # Note: Google for "python arbitrary arguments" and look for "*args"
 
 # YOUR CODE HERE
+def f2(*args):
+    sum = 0
+    for x in args:
+        sum += x
+    return sum
```
Comment (birdcar) [inline src/11_args.py:22]: Here's a fun one. Your code is absolutely accurate, but there's actually a built-in `sum()` function that takes any iterable or iterator. Since `*args` takes every argument and packs it into a tuple, we can pass the tuple directly to `sum` like so:

```python
def f2(*args):
    return sum(args)
```

File: src/11_args.py [hunk 11]
```diff
@@ -22,14 +29,19 @@
...
 # it returns that value plus 1. If two arguments, it returns the sum of the
 # arguments.
 # Note: Google "python default arguments" for a hint.
 
 # YOUR CODE HERE
+def f3(a, b=None):
+    if b is None:
+        return a + 1
+    else:
+        return a + b
```
Comment (birdcar) [inline src/11_args.py]: First off, great use of a non-mutable default argument. Due to how Python's object / reference system works, you never ever want to use a mutable object (like a list) as a default argument.

with that in mind, lets take that turnery expression we used earlier and clean this up a little bit to shorten our code and make our meaning a little more like plain english:

```python
def f3(a, b=None):
    return a + 1 if b is None else a + b
```

---

**Note:** the `if b is None` part of this is actually very important. If we were to shorten this further to be `if b` like so:

```python
def f3(a, b=None):
    return a + 1 if b else a + b
```

And then someone tried to call our function with `0` as the second argument (i.e. `f3(1, 0)`), our function would actually return `2` because `b` would evaluate to `False`. Likely _not_ what our function caller intended.

File: src/13_file_io.py [hunk 12]
```diff
@@ -10,10 +10,25 @@
 # Note: pay close attention to your current directory when trying to open "foo.txt"
 
 # YOUR CODE HERE
+with open('foo.txt') as f1:
+    read_file1 = f1.read()
+    print(read_file1)
```
Comment (birdcar) [inline src/13_file_io.py]: I wanted to take a moment to expound on the note in line 10.

When you execute a Python program, if you don't specify an absolute path in a function like "open", it assumes that you want a relative path based on what your current working directory is when you executed the program.

This means that if you go one directory _above_ `src` to the root directory of the repo, and run `python src/13_file_io.py` you'll actually get an error, because there is no `foo.txt` file in the root directory of the repository.

Since working with files is a hella common thing in Python, lets chat about how we could make this less fragile.

Python provides you with some "dunder" attributes when you're inside of a module (`.py` file). One of these is `__file__`, which is populated by the Python runtime when it executes a module, and it is populated with the relative path from the module caller's location to the actual module currently being executed.

What this means practically is that if we're in the `src` directory and we run `python 13_file_io.py` then `__file__` will equal `'13_file_io.py`. But if we're in the root of the repository it'll equal `src/13_file_io.py`.

We can use this to our advantage in a couple ways, but the _best_ way in my opinion would be to use [the Path object from the `pathlib` module](https://docs.python.org/3/library/pathlib.html). I'll let you read the docs to learn it in detail, but the code below will be a good pattern to remember for future projects (*especailly* in stuff like Django and Flask):

```python
from pathlib import Path

# The Path call will generate a platform agnostic (i.e. Windows users won't mess us up) path to `__file__`
#
# The resolve method will then convert that path from relative to absolute, providing a path that will 
# _always_ point to the file no matter where the file is being run.
#
# Finally, the `parent` attribute will return a new Path object pointed to the directory that the file is 
# in, in this case the `src` directory
PARENT_DIR = Path(__file__).resolve().parent

# We can then build paths from the parent directory of the file by using slashes
with open(PARENT_DIR/'foo.txt') as f1:
    read_file1 = f1.read()
    print(read_file1)
```

It'll solve all of that nasty, fragile "be careful where you run the file from!!!" warnings.

**Review by birdcar — COMMENTED** [review]
Hey @thisbenrogers 👋 

Nothing you did is wrong (well done), this is just a friendly Pythonista saying "Hey! Look at this cool stuff!"

Go ahead and delete the comments whenever you want, they're not necessary at all. 🍻

File: src/08_comprehensions.py (hunk 3 above)
Comment (thisbenrogers) [reply src/08_comprehensions.py]: Lol
Look at line 15 and laugh with current-me at past-me, creating problems for future-me.

File: src/09_dictionaries.py (hunk 7 above)
Comment (thisbenrogers) [reply src/09_dictionaries.py:58]: Good grief I've done this before too! Check line 13 of 03_modules.py
